discojs/src/types.ts

```diff
@@ -1,11 +1,38 @@
-import { Dataset, Image, Tabular, Text } from "./dataset/index.js"
+import { List } from "immutable";
 
-export type TypedDataset =
-  | ["image", Dataset<Image>]
-  | ["tabular", Dataset<Tabular>]
-  | ["text", Dataset<Text>];
+import type { Image, processing, Tabular, Text } from "./index.js";
 
-export type TypedLabeledDataset =
-  | ["image", Dataset<[Image, label: string]>]
-  | ["tabular", Dataset<Tabular>]
-  | ["text", Dataset<Text>];
+/**
+ * The data that we handle goes through various stages.
+ * The labels also gets transformed at each stage.
+ *
+ * raw -> model encoded -> inferred
```

Comment:
> but it seems to be representing a dataset ready to be fed to a model, is this right?

yeah, pretty much. `ModelEncoded` is here both the input and the output of the model: the two are needed for training, `predict` takes the first one and return the second, validator might only need the input, …

> What about changing the names to RawDataset -> ModelInputDataset -> ModelOutputDataset?

I was using previously smthDataset, but it began exploding when dealing with Batched and/or WithLabel; also, it is not a Dataset in itself but a type of data _inside_ a Dataset.
instead of having it as a ~root type, I can push it into a module, as to have `data.Raw`, `data.ModelEncoded`, … I don't like the name "data" as it is quite undescriptive, maybe you've a better one?

Reply:
Maybe rename data to DatasetFormat? and have Raw, Training, Prediction and Validator formats?